src/Microsoft.AspNetCore.Routing/Matching/EndpointComparer.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    // Use to sort and group Endpoints. RouteEndpoints are sorted before other implementations.
+    //
+    // NOTE:
+    // When ordering endpoints, we compare the route templates as an absolute last resort.
+    // This is used as a factor to ensure that we always have a predictable ordering
+    // for tests, errors, etc.
+    //
+    // When we group endpoints we don't consider the route template, because we're trying
+    // to group endpoints not separate them.
+    //
+    // TLDR: 
+    //  IComparer implementation considers the template string as a tiebreaker.
+    //  IEqualityComparer implementation does not.
+    //  This is cool and good.
+    internal class EndpointComparer : IComparer<Endpoint>, IEqualityComparer<Endpoint>
+    {
+        private readonly IComparer<Endpoint>[] _comparers;
+
+        public EndpointComparer(IEndpointComparerPolicy[] policies)
+        {
+            // Order, Precedence, (others)...
+            _comparers = new IComparer<Endpoint>[2 + policies.Length];
+            _comparers[0] = OrderComparer.Instance;
+            _comparers[1] = PrecedenceComparer.Instance;
+            for (var i = 0; i < policies.Length; i++)
+            {
+                _comparers[i + 2] = policies[i].Comparer;
+            }
+        }
+
+        public int Compare(Endpoint x, Endpoint y)
+        {
+            // We don't expose this publicly, and we should never call it on
+            // a null endpoint.
+            Debug.Assert(x != null);
+            Debug.Assert(y != null);
+
+            var compare = CompareCore(x, y);
+
+            // Since we're sorting, use the route template as a last resort.
+            return compare == 0 ? ComparePattern(x, y) : compare;
+        }
+
+        private int ComparePattern(Endpoint x, Endpoint y)
```

Comment:
Worth a comment I think that a RouteEndpoint always comes before a non-Route-endpoint in the rules of this method.